tcos-standalone/tcos-standalone.1
.TH tcos-standalone 1 "Nov 29, 2007" "tcos-standalone man page"

.SH NAME
tcos-standalone \- Python GTK2 gui that allow to enable or disable 
tcos-standalone mode.

.SH DESCRIPTION

.PP
\fBtcos-standalone\fP is a small python GUI  to enable or disable 
tcos-standalone mode


tcos-standalone edit /etc/default/tcos-standalone and creates /etc/tcospasswd

.SH OPTIONS
.B [no options]
 With no options will run in normal mode.

.B --help
 Some usage help about tcos-standalone

.B --debug
 Show verbose output when running

.SH BUGS
Contact Mario Izquierdo <mariodebian@gmail.com>

Homepage: http://www.tcosproject.org

.SH AUTHOR
Mario Izquierdo (mariodebian) <mariodebian@gmail.com>


debian/changelog
initramfs-tools-tcos (0.89.53) unstable; urgency=low

  * move some tcos-bottom script later
  * scripts/tcos-bottom/80shell: stop bootchart if exists
  * scripts/tcos-bottom/57devices: test ltspfsd
  * scripts/tcos-bottom/55mount_listener: only start if USB and IDE is
    enabled
  * set chmod +x in scripts
  * bin/tcos-genbootchart: use DISPLAY if not arg is passed
  * set 2.6.32-trunk-486 for Debian unstable
  * conf/base.conf: add TCOS_USBIP new var, disabled by default
  * hooks-addons/19usbip: use new var TCOS_USBIP
  * debian/tcos-standalone.init: provides tcos-standalone
  * debian/copyright: replace (C) with ©
  * debian/control: add misc:Depends and fix typo
  * xmlrpc/man/update-tcospasswd.1 tcos-standalone/tcos-standalone.1
    bin/tcos-gdm-autologin.1 bin/tcos-buildchroot.1 bin/gentcos.1: split
    lines longer than 80 chars
  * Makefile conf/tcos.conf.light: Add new light template run in 46 MB
  * conf/tcos-run-functions.sh: update list of killed apps
  * tcos-standalone/tcos-standalone.desktop: remove obsolete Encoding
  * New XMLRPC method: config, read vars from tcos.conf in thin client

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Fri, 18 Dec 2009 19:04:14 +0100

initramfs-tools-tcos (0.89.52) unstable; urgency=low

  * Enable/disable composite Xorg extension
  * fix if::else TCOS_COMPOSITE
  * downgrade some depends to recommends
  * cosmetic hanges in standalone init script
  * bin/tcos-buildchroot: copy /etc/localtime /etc/adjtime if exists
  * hooks-addons/50xorg: copy linked libs
  * add TCOS_CPUFREQ support with powersave scaling governor
  * tcos-standalone/tcos-standalone.py replace obsolete popen2 with
    subprocess
  * scripts/tcos-bottom/30devices: disable -a in ltspfsd, not needed
    anymore
  * remove rtptools and sox depends, audio multicast use pulseaudio
    native RTP module
  * fix typo suggest => suggests

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun, 13 Dec 2009 19:24:56 +0100

initramfs-tools-tcos (0.89.51) unstable; urgency=low

  * new bin waitforX, to e used in startup scripts
  * copy waitforX and use it instead of 'pidof Xorg'
  * dnsmasq by default
  * detect VGA vendor with new script get_vga_vendor
  * update USBIP bins PATH
  * add 2 new openchrome cards
  * hooks-addons/53mga: fix typo
  * hooks-addons/21pulseaudio: fix typo
  * tcos-genbootchart: move to /usr/bin
  * hooks-addons/05devices call ltspfs-set-xprop in X11/PreRun
  * Update copyright
  * fix path in tcos-genbootchart

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 07 Dec 2009 21:37:52 +0100

initramfs-tools-tcos (0.89.50) unstable; urgency=low

  * Depend on ltspfsd-core and ltspfs
  * Use evdev as keyboard driver when kbd is not found
  * put XbdModel=abnt2 in evdev driver
  * update depends, change to i386 the thin client depends
  * update kernels
  * revert arch depends

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun, 22 Nov 2009 20:14:37 +0100

initramfs-tools-tcos (0.89.49) unstable; urgency=low

  * remove libnss-db, new libc conflicts with it
  * set home if getent fails
  * bin/tcos-buildchroot: remove libnss-db from PACKAGES
  * hooks-addons/00main: copy libnss_db.so.2 if exists
  * bin/get_xauth: read cookie from GDM/KDM when no user is logged
  * scripts/tcos-bottom/10ldconfig: export env vars in /root/.profile
  * scripts/tcos-bottom/80shell: change order in PS1 var
  * debian/initramfs-tools-tcos.postinst: fix typo

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun, 08 Nov 2009 14:57:29 +0100

initramfs-tools-tcos (0.89.48) unstable; urgency=low

  * fix startscale STANDALONE auth argument
  * copy localtime and adjtime from system to chroot
  * save date in BIOS after run ntpdate and copy /etc/services always,
    not only if inetd is ENABLED
  * bin/get_filesystem: call udevadm if exists
  * bin/tryXorg.sh: fix AutoAddDevices in xorg.conf.auto

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 13 Oct 2009 12:02:21 +0200

initramfs-tools-tcos (0.89.47) unstable; urgency=low

  * fix typo and use DPATH inseat of reserved var PATH
  * udev/mount_listener.c don't show errors if poll() return 0
  * hooks-addons/03udev: copy /lib/udev/rules.d/40-alsa.rules, 
    Ubuntu Jaunty fix sound
  * add support to update BIOS date/time with ntpdate, disabled by default
  * use blkid, from util-linux (if avalaible) to get filesystem type
  * call user scripts in /etc/tcos/pxelinux.d/*.hook files
  * log ntpdate in initramfs.debug file
  * fix typos in example install
  * add new files to initramfs package
  * check for reopen in x11vnc before using it
  * fix exec pxelinux.d hooks and fix example hack

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun, 04 Oct 2009 15:43:59 +0200

initramfs-tools-tcos (0.89.46) unstable; urgency=low

  * pulseaudio: set exit-idle-time to -1, only set SHM if low memory <
    64MiB, and don't copy udev or x11 modules
  * create p910nd lock dir
  * update Ubuntu Jaunty kernel to 2.6.28-15-generic
  * don't run vol_id if not found, and add udevinfo detection method
  * udev/tcos-udevd.sh udev/save-udev.sh: translate DEVPATH to old short
    one

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Thu, 01 Oct 2009 16:03:36 +0200

initramfs-tools-tcos (0.89.45) unstable; urgency=low

  * Allow to start x11vnc server in scale mode
  * Include dbus utils, disabled by default
  * PulseAudio revert to deprecated module-detect add users and revert
    module-x11-publish => pax11publish
  * Remove chroot package from DISTRO_PACKAGES

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 30 Sep 2009 22:41:29 +0200

initramfs-tools-tcos (0.89.44) unstable; urgency=low

  * Use abnt2 keymap for Brasil users
  * Add support for NTFS-3G
  * Add more info about dnsmasq and gPXE
  * Allow to personalize usplash resolution
  * Fix rdesktop IP detection and use correct keymap
  * xmlrpc/*:
    - Add support to get screensize in info.[ch]


tcos-standalone/tcos-standalone.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
##########################################################################
# TcosStandalone writen by MarioDebian <mariodebian@gmail.com>
#
#    TcosStandalone version __VERSION__
#
# Copyright (c) 2007 Mario Izquierdo <mariodebian@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
###########################################################################

import sys
import os

import pygtk
pygtk.require('2.0')
from gtk import *
import gtk.glade

from time import time, sleep
import getopt
from gettext import gettext as _
from gettext import bindtextdomain, textdomain
from locale import setlocale, LC_ALL

from subprocess import Popen, PIPE, STDOUT
#import popen2
from threading import Thread

#import threading
gtk.gdk.threads_init()

import gobject

debug=False
PACKAGE="tcos-standalone"

tcos_standalone_file="/etc/default/tcos-standalone"
tcos_standalone_pass="/etc/tcospasswd"
daemon_cmd="/etc/init.d/tcos-standalone"

# if exec from svn or sources dir
if os.path.isdir('../debian'):
    LOCALE_DIR = "./po/"
    GLADE_DIR = "./"
    IMG_DIR = "./images/"
    print "exec in sources dir"
else:
    GLADE_DIR = "/usr/share/tcos-standalone/"
    IMG_DIR = "/usr/share/tcos-standalone/images/"
    LOCALE_DIR = "/usr/share/locale"

def print_debug(txt):
    if debug:
        print "%s::%s" %("tcos-standalone", txt)
    return

def usage():
    print "tcos-standalone help:"
    print ""
    print "   tcos-standalone -d [--debug]  (write debug data to stdout)"
    print "   tcos-standalone -h [--help]   (this help)"


try:
    opts, args = getopt.getopt(sys.argv[1:], ":hd", ["help", "debug"])
except getopt.error, msg:
    print msg
    print "for command line options use tcosconfig --help"
    sys.exit(2)

# process options
for o, a in opts:
    if o in ("-d", "--debug"):
        print "DEBUG ACTIVE"
        debug = True
    if o in ("-h", "--help"):
        usage()
        sys.exit()


class TcosStandalone:
    def __init__(self):
        print_debug("__init__()")
        
        # vars
        self.v={}
        self.v["daemon_enabled"]=None
        self.v["daemon_running"]={}
        self.v["username"]=None
        self.v["password"]=None
        self.v["newconfig"]=False
        self.v["newpass"]=False
        
        gtk.glade.bindtextdomain(PACKAGE, LOCALE_DIR)
        gtk.glade.textdomain(PACKAGE)

        # gettext support
        setlocale( LC_ALL )
        bindtextdomain( PACKAGE, LOCALE_DIR )
        textdomain( PACKAGE )
        
        # Widgets
        self.ui = gtk.glade.XML(GLADE_DIR + 'tcos-standalone.glade')
        self.mainwindow = self.ui.get_widget('mainwindow')
        self.mainwindow.set_icon_from_file(IMG_DIR +'tcos-icon-32x32.png')
        
        # close windows signals
        self.mainwindow.connect('destroy', self.quitapp )
        self.mainwindow.connect("delete_event", self.quitapp)
        
        # read vars
        self.v["daemon_enabled"]=self.read_conf(tcos_standalone_file, "TCOS_STANDALONE_START", 1, '=')
        self.v["ssl_enabled"]=self.read_conf(tcos_standalone_file, "TCOS_XMLRPC_SSL", 0, '=')
        self.v["username"]=self.read_conf(tcos_standalone_pass, "root", 0, ':')
        daemons=self.exe_cmd(daemon_cmd + " status")
        self.v['daemon_running'][daemons.split(' ')[0]]=daemons.split(' ')[1]
        
        #for daemon in daemons:
        #    print_debug("daemon %s"%daemon)
        #    #self.v["daemon_running"][daemon.split(' ')[0]]=daemon.split(' ')[1]
        
        # widgets
        self.w={}
        self.w["txt_username"] = self.ui.get_widget('txt_username')
        self.w["txt_password"] = self.ui.get_widget('txt_password')
        self.w["lb_status"] = self.ui.get_widget('lb_status')
        self.w["ck_standalone_start"] = self.ui.get_widget('ck_standalone_start')
        self.w["ck_enable_ssl"] = self.ui.get_widget('ck_enable_ssl')
        self.w["lb_tcosxmlrpc"] = self.ui.get_widget('lb_tcosxmlrpc')
        #self.w["lb_busybox"] = self.ui.get_widget('lb_busybox')
        self.w["img_tcosxmlrpc"] = self.ui.get_widget('img_tcosxmlrpc')
        #self.w["img_busybox"] = self.ui.get_widget('img_busybox')
        self.w["bt_apply"] = self.ui.get_widget('bt_apply')
        self.w["bt_stop"] = self.ui.get_widget('bt_stop')
        self.w["bt_start"] = self.ui.get_widget('bt_start')
        
        self.w["ck_standalone_start"].connect('toggled', self.ck_change )
        self.w["ck_enable_ssl"].connect('toggled', self.ck_change )
        
        if self.v["daemon_enabled"] == '1':
            self.w["ck_standalone_start"].set_active(True)
        else:
            self.w["ck_standalone_start"].set_active(False)
        
        if self.v["ssl_enabled"] == '1':
            self.w["ck_enable_ssl"].set_active(True)
        else:
            self.w["ck_enable_ssl"].set_active(False)
        
        
        self.w["txt_username"].connect('changed', self.pass_change )
        self.w["txt_password"].connect('changed', self.pass_change )
        
        if not self.v["username"]:
            self.w["lb_status"].set_text( _("Not user and password configured.\nUsing root system user.") )
        else:
            self.w["txt_username"].set_text(self.v["username"])
        
        
        self.w["bt_apply"].connect('clicked', self.apply_changes)
        
        self.w["bt_stop"].connect('clicked', self.stop_daemon)
        self.w["bt_start"].connect('clicked', self.start_daemon)

        print_debug ( self.v )

        self.v["newconfig"]=False
        self.v["newpass"]=False
        self.update_images()
        self.refresh_buttons()
        

    def ck_change(self, *args):
        self.v["newconfig"]=True
        self.refresh_buttons()

    def pass_change(self, *args):
        self.v["newpass"]=True
        self.refresh_buttons()


    def refresh_buttons(self):
        print_debug("refresh_buttons() newconfig=%s newpass=%s" %(self.v["newconfig"], self.v["newpass"]))
        if self.v["newconfig"] or self.v["newpass"]:
            # configure buttons
            self.w["bt_apply"].set_sensitive(True)
        else:
            # configure buttons
            self.w["bt_apply"].set_sensitive(False)
            

    def apply_changes(self, *args):
        if self.v["newconfig"]:
            print_debug("apply_changes() newconfig")
            # daemon enabled when open app
            
            if self.w["ck_enable_ssl"].get_active() != 1:
                self.save_conf(tcos_standalone_file, "TCOS_XMLRPC_SSL", 0, '=')
                if self.v["daemon_enabled"] != '0':
                    self.exe_cmd(daemon_cmd + " stop")
                    self.exe_cmd(daemon_cmd + " start")
                print_debug("apply_changes() disabling SSL")
            else:
                self.save_conf(tcos_standalone_file, "TCOS_XMLRPC_SSL", 1, '=')
                if self.v["daemon_enabled"] != '0':
                    self.exe_cmd(daemon_cmd + " stop")
                    self.exe_cmd(daemon_cmd + " start")
                print_debug("apply_changes() enabling SSL")
                
            if self.v["daemon_enabled"] != '0':
                if self.w["ck_standalone_start"].get_active() != 1:
                    # disable daemon
                    self.exe_cmd(daemon_cmd + " stop")
                    self.save_conf(tcos_standalone_file, "TCOS_STANDALONE_START", 0, '=')
                    self.v["daemon_enabled"]='0'
            # daemon disabled when open app
            else:
                if self.w["ck_standalone_start"].get_active() == 1:
                    # enable daemon
                    self.save_conf(tcos_standalone_file, "TCOS_STANDALONE_START", 1, '=')
                    self.exe_cmd(daemon_cmd + " stop")
                    self.exe_cmd(daemon_cmd + " start")
                    self.v["daemon_enabled"]='1'
            self.v["newconfig"]=False
        
        if self.v["newpass"]:
            tmp=self.exe_cmd("/usr/sbin/tcospasswd %s" %(self.w["txt_password"].get_text() ) )
            fw=file(tcos_standalone_pass, 'w')
            fw.write("%s:%s\n" %(self.w["txt_username"].get_text(), tmp) )
            fw.close()
            self.w["lb_status"].set_text( _("New username and password used.\n") )
            self.v["newpass"]=False
        
        self.refresh_buttons()
        self.update_images()

    def stop_daemon(self, *args):
        self.exe_cmd(daemon_cmd + " stop")
        self.update_images()

    def start_daemon(self, *args):
        self.exe_cmd(daemon_cmd + " start")
        self.update_images()
        

    def update_images(self):
        daemons=self.exe_cmd(daemon_cmd + " status")
        self.v['daemon_running'][daemons.split(' ')[0]]=daemons.split(' ')[1]
        #for daemon in daemons:
        #    self.v["daemon_running"][daemon.split(' ')[0]]=daemon.split(' ')[1]
            
        if self.v["daemon_running"]["tcosxmlrpc"] == "running":
            self.w["img_tcosxmlrpc"].set_from_stock(gtk.STOCK_YES, 4)
            self.w["lb_tcosxmlrpc"].set_text( _("tcosxmlrpc running") )
            self.w["bt_stop"].set_sensitive(True)
            self.w["bt_start"].set_sensitive(False)
        else:
            self.w["img_tcosxmlrpc"].set_from_stock(gtk.STOCK_NO, 4)
            self.w["lb_tcosxmlrpc"].set_text( _("tcosxmlrpc stopped") )
            self.w["bt_stop"].set_sensitive(False)
            self.w["bt_start"].set_sensitive(True)
        
        #if self.v["daemon_running"]["busybox"] == "running":
        #    self.w["img_busybox"].set_from_stock(gtk.STOCK_YES, 4)
        #    self.w["lb_busybox"].set_text( _("busybox running") )
        #else:
        #    self.w["img_busybox"].set_from_stock(gtk.STOCK_NO, 4)
        #    self.w["lb_busybox"].set_text( _("busybox stopped") )


    def read_conf(self, fname, fvar, item=0, sep='='):
        if not os.path.isfile(fname): return None
        f=open(fname, 'r')
        data=f.readlines()
        f.close()
        for line in data:
            if line.startswith(fvar):
                return line.replace('\n','').split(sep)[item]

    def save_conf(self, fname, fvar, value, sep='='):
        if not os.path.isfile(fname): return None
        f=open(fname, 'r')
        data=f.readlines()
        f.close()
        print_debug("save_conf() fname=%s fvar=%s value=%s" %(fname, fvar, value))
        fw=file(fname, 'w')
        for i in range(len(data)):
            if data[i].startswith(fvar):
                print_debug("save_conf() line=%s%s%s" %(fvar, sep, value) )
                fw.write( "%s%s%s\n" %(fvar, sep, value) )
            else:
                fw.write(data[i])
                print_debug("save_conf() line=%s" %(data[i].replace('\n','')) )
        fw.close()
        
    def exe_cmd(self, cmd, verbose=1):
        print_debug("exe_cmd() cmd=%s" %cmd)
        output=[]
        p=Popen(cmd, shell=True, bufsize=0, stdout=PIPE, stderr=STDOUT, close_fds=True)
        stdout = p.stdout
        #(stdout, stdin) = popen2.popen2(cmd)
        #stdin.close()
        for line in stdout:
            if line != '\n':
                line=line.replace('\n', '')
                output.append(line)
        if len(output) == 1:
            return output[0]
        elif len(output) > 1:
            if verbose==1:
                print_debug ( "exe_cmd(%s) %s" %(cmd, output) )
            return output
        else:
            if verbose == 1:
                print_debug ( "exe_cmd(%s)=None" %(cmd) )
            return []


    def quitapp(self,*args):
        print_debug ( "Exiting" )
        self.mainloop.quit()

    def run (self):
        self.mainloop = gobject.MainLoop()
        try:
            self.mainloop.run()
        except KeyboardInterrupt: # Press Ctrl+C
            self.quitapp()
   


if __name__ == '__main__':
    app = TcosStandalone ()
    # Run app
    app.run ()
